# **WOOF_Project**

Scripts and pipelines used to analyze extant and ancient HTS data from the WOOF project. The steps envolve pre-processing, mapping, variant calling and population genomics analysis. Designed for the analysis of Illumina sequences obtained from Portuguese and Spanish specimens generated by the WOOF project and reference sequences retrieved from database.

## **List of tools/software needed**

| Program | Version | Source | Application |
|----------|----------|----------|----------|
| FastQC | | | Read quality control |
| Cutadapt | 1.18 | | Read pre-processing |
| BWA | | https://github.com/lh3/bwa | Read alignment |
| Picard | | | Post-alignment processing |
| GATK | | https://software.broadinstitute.org/gatk | Post-alignment processing + SNP calling |
| Qualimap | | http://qualimap.bioinfo.cipf.es/ | Alignment quality control |
| SAMtools | | http://www.htslib.org/ | BAM file manipulation |
| ANGSD | | http://www.popgen.dk/angsd | Estimation of genotype likelihoods + Population analysis |
| Admixture | | | Admixture analysis |
| PLINK | | | VCF file manipulation + PCA |
| BCFtools | | http://www.htslib.org/ | VCF file manipulation |
| VCFtools | | | VCF file manipulation |
| MUSCLE | | http://www.drive5.com/muscle/ | Mitochondrial phylogeny |
| RAxML | | https://github.com/stamatak/standard-RAxML | Mitochondrial phylogeny |
| R | | https://cran.r-project.org/bin/windows/ | Plotting + Statistical analysis |
| Treemix | | | Population splits and admixture analysis |
| Figtree | | http://tree.bio.ed.ac.uk/software/figtree/| |


## Fastq2BAM - Read filter and mapping
- Scripts to analyze modern HTS data from preprocessing to mapping.
- (SCRIPTS_IN_PROGRESS) has SCRIPTS being developed to improve future analysis.
- The genome reference is based on the Boxer assembly from XXX

## FILO - Mitochondrial Filogenies

## POP_GEN - Population analyses based on variant calling

## Chalcolithic_Paper
Scripts and pipelines used to analyze modern and acient HTS data for the paper "A multidisciplinary study of Iberian Chalcolithic dogs" published in JAS:Reports.
Access the paper at https://doi.org/10.1016/j.jasrep.2022.103338.

