## **List of tools/software needed**

| Program | Version | Source | Application |
|----------|----------|----------|----------|
| FastQC | | | Read quality control |
| Cutadapt | | | Read pre-processing |
| BWA | | https://github.com/lh3/bwa | Read alignment |
| Picard | | | Post-alignment processing |
| GATK | | https://software.broadinstitute.org/gatk | Post-alignment processing + SNP calling |
| Qualimap | | http://qualimap.bioinfo.cipf.es/ | Alignment quality control |
| SAMtools | | http://www.htslib.org/ | BAM file manipulation |
| ANGSD | | http://www.popgen.dk/angsd | Estimation of genotype likelihoods + Population analysis |
| Admixture | | | Admixture analysis |
| PLINK | | | VCF file manipulation + PCA |
| BCFtools | | http://www.htslib.org/ | VCF file manipulation |
| VCFtools | | | VCF file manipulation |
| MUSCLE | | | Mitochondrial phylogeny |
| RAxML | | https://github.com/stamatak/standard-RAxML | Mitochondrial phylogeny |
| R | | | Plotting + Statistical analysis |
| Treemix | | | Population splits and admixture analysis |
| Figtree | | http://tree.bio.ed.ac.uk/software/figtree/| |

The genome reference is based on the Boxer assembly from XXX
